src/java.net.http/share/classes/jdk/internal/net/http/Http3Connection.java

```diff
@@ -873,6 +888,44 @@ boolean tryReserveForPoolCheckout() {
         }
     }
 
+    /**
+     * Mark this connection as being present or absent from the connection pool.
+     */
+    void setPooled(final boolean present) {
+        this.presentInConnPool = present;
+    }
+
+    /**
+     * This callback method is invoked by the QUIC layer when it notices that this
+     * connection hasn't seen any traffic for certain period of time. QUIC
+     * invokes this method to ask HTTP/3 whether the QUIC layer
+     * should generate traffic to keep this connection active.
+     * This method returns true, indicating that the traffic must be generated,
+     * if this HTTP/3 connection is in pool and there's no current request/response
+     * in progress over this connection (i.e. the HTTP/3 connection is idle in the
+     * pool waiting for any new requests to be issued by the application).
+     */
+    private boolean isQUICTrafficGenerationRequired() {
+        if (!isOpen()) {
+            return false;
+        }
+        lock();
+        try {
+            // if there's no HTTP/3 request/responses in progress and the connection is
+            // in the pool (thus idle), then we instruct QUIC to generate traffic on the
+            // QUIC connection to prevent it from being idle terminated.
+            final boolean generateTraffic = this.presentInConnPool
```

Comment:
I wonder if `this.presentInConnPool` can be replaced with `!this.finalStream`

Reply:
Me and Daniel discussed about this and we decided to include a `!finalStream` check too. The PR has been updated with this change. The test continues to pass and I've triggered a CI run with these latest updates.